File: ios/GradeThread/Marketplaces/BulkPricing/BulkPricingService.swift
import Foundation

/// One active eBay listing eligible for a bulk price/quantity edit.
struct BulkListing: Identifiable, Equatable {
    let id: String
    let title: String
    let price: Double
    let quantity: Int?
}

/// One update sent to the bulk endpoint. Encoded via EdgeAPI's convertToSnakeCase
/// encoder (listingId -> listing_id). nil price/quantity are omitted.
struct BulkPriceQtyUpdate: Encodable, Equatable {
    let listingId: String
    let price: Double?
    let quantity: Int?
}

struct BulkPriceQtyResult: Decodable, Equatable {
    let listingId: String
    let ok: Bool
    let error: String?
}

struct BulkPriceQtyResponse: Decodable {
    let results: [BulkPriceQtyResult]
    let succeeded: Int
    let total: Int
}

/// Data layer for bulk price/quantity (US-1046 clean surface). Reads `listings`
/// directly through supabase-swift (RLS scopes to the caller via inventory-item
/// ownership) and applies updates via the consolidated edge endpoint, which
/// pushes to eBay's bulk_update_price_quantity. Behind a protocol so the store
/// is testable.
protocol BulkPricingProviding {
    func listings() async throws -> [BulkListing]
    func apply(updates: [BulkPriceQtyUpdate]) async throws -> BulkPriceQtyResponse
}

struct BulkPricingService: BulkPricingProviding {
    /// Wire row from `listings`. Price/quantity decode leniently because
    /// PostgREST can serialize numeric columns as JSON numbers OR strings.
    private struct Row: Decodable {
        let id: String
        let title: String?
        let price: Double
        let quantity: Int?
        let offerId: String?

        private enum CodingKeys: String, CodingKey {
            case id
            case title = "listing_title"
            case price = "listing_price"
            case quantity
            case offerId = "platform_offer_id"
        }

        init(from decoder: Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            id = try c.decode(String.self, forKey: .id)
            title = try c.decodeIfPresent(String.self, forKey: .title)
            offerId = try c.decodeIfPresent(String.self, forKey: .offerId)
            if let d = try? c.decode(Double.self, forKey: .price) {
                price = d
            } else if let s = try? c.decode(String.self, forKey: .price), let d = Double(s) {
                price = d
            } else {
                price = 0
            }
            if let i = try? c.decodeIfPresent(Int.self, forKey: .quantity) {
                quantity = i
            } else if let s = try? c.decodeIfPresent(String.self, forKey: .quantity) {
                quantity = s.flatMap(Int.init)
            } else {
                quantity = nil
            }
        }
    }

    func listings() async throws -> [BulkListing] {
        let response = try await SupabaseShared.client
            .from("listings")
            .select("id, listing_title, listing_price, quantity, platform_offer_id")
            .eq("platform", value: "ebay")
            .eq("listing_status", value: "active")
            .order("listed_at", ascending: false)
            .limit(500)
            .execute()
        let rows = (try? JSONDecoder().decode([Row].self, from: response.data)) ?? []
        // Only offers can be bulk-updated; filter client-side to avoid an
        // is-not-null filter quirk.
        return rows.compactMap { row in
            guard let offer = row.offerId, !offer.isEmpty else { return nil }
            return BulkListing(
                id: row.id,
                title: (row.title?.isEmpty == false ? row.title! : "Untitled listing"),
                price: row.price,
                quantity: row.quantity
            )
        }
    }

    func apply(updates: [BulkPriceQtyUpdate]) async throws -> BulkPriceQtyResponse {
        struct Body: Encodable { let updates: [BulkPriceQtyUpdate] }
        return try await EdgeAPI.shared.postJSON(
            "/api/flipdesk/ebay/listings/bulk-price-quantity",
            body: Body(updates: updates)
        )
    }
}

File: android/app/src/main/java/com/gradethread/app/marketplaces/pricing/BulkPricingService.kt
package com.gradethread.app.marketplaces.pricing

import com.gradethread.app.platform.net.EdgeApi
import com.gradethread.app.platform.net.EdgeApiError
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.serialization.Serializable
import javax.inject.Inject
import javax.inject.Named
import javax.inject.Singleton

/**
 * US-1355: data for the bulk price editor.
 *
 * Listings are read straight from Postgres rather than the Room mirror, for one
 * reason: the editor needs `listing_title`, which the local cache doesn't
 * carry. Reads go through the anon client, so RLS scopes them to the owner.
 * The push goes through the edge, which owns the eBay call.
 */
@Singleton
class BulkPricingService @Inject constructor(
    @Named("shared") private val edge: EdgeApi,
    private val client: SupabaseClient,
) {

    @Serializable
    private data class BulkRequest(val updates: List<BulkPriceUpdate>)

    companion object {
        const val BULK_PATH = "/api/flipdesk/ebay/listings/bulk-price-quantity"
        private const val TABLE = "listings"
        private const val PAGE = 500
    }

    /**
     * Active eBay listings that can actually be repriced.
     *
     * Filtered to rows with an offer id: the bulk endpoint addresses eBay's
     * Inventory API, which only knows about offers GradeThread published. An
     * imported listing has no offer, so including it would put rows in the
     * editor that always fail. Done client-side to avoid the not-null filter
     * quirk on the hosted PostgREST.
     */
    suspend fun listings(): List<BulkListing> =
        client.from(TABLE).select(
            Columns.raw("id, listing_title, listing_price, quantity, platform_offer_id"),
        ) {
            filter {
                eq("platform", "ebay")
                eq("listing_status", "active")
            }
            order("listed_at", Order.DESCENDING)
            limit(PAGE.toLong())
        }.decodeList<BulkListingRow>()
            .filter { !it.platformOfferId.isNullOrBlank() }
            .map {
                BulkListing(
                    id = it.id,
                    title = it.listingTitle?.takeIf { t -> t.isNotBlank() } ?: "Untitled listing",
                    price = it.listingPrice,
                    quantity = it.quantity,
                )
            }

    /** Push the updates. The server reports per-listing, never all-or-nothing. */
    suspend fun apply(updates: List<BulkPriceUpdate>): BulkPriceResponse =
        edge.json.decodeFromString(
            BulkPriceResponse.serializer(),
            edge.postRaw(
                BULK_PATH,
                edge.json.encodeToString(BulkRequest.serializer(), BulkRequest(updates)),
            ),
        )

    fun message(error: Throwable): String =
        (error as? EdgeApiError)?.userMessage()
            ?: error.message
            ?: "Couldn't push those prices to eBay."
}
